c7n/resources/servicecatalog.py

```diff
@@ -0,0 +1,154 @@
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+          - name: service-catalog-portfolio-delete
+            resource: aws.catalog-portfolio
+            filters:
+              - type: cross-account
+            actions:
+              - type: remove-shared-accounts
+                accounts: matched
+              - delete
+    """
+
+    schema = type_schema('delete')
+    permissions = ('servicecatalog:DeletePortfolio',)
+
+    def process(self, portfolios):
+        client = local_session(self.manager.session_factory).client('servicecatalog')
+        for r in portfolios:
+            self.manager.retry(client.delete_portfolio, Id=r['Id'], ignore_err_codes=(
+                'ResourceNotFoundException',))
+
+
+@CatalogPortfolio.filter_registry.register('cross-account')
+class CatalogPortfolioCrossAccount(CrossAccountAccessFilter):
+    """Check for account ids that the service catalog portfolio is shared with
+    """
+
+    schema = type_schema(
+        'cross-account',
+        whitelist_accounts_from={'$ref': '#/definitions/filters_common/value_from'},
+        whitelist={'type': 'array', 'items': {'type': 'string'}})
+
+    permissions = ('servicecatalog:ListPortfolioAccess',)
+    annotation_key = 'c7n:CrossAccountViolations'
+
+    def violation_checker(self, client, accounts, resources):
+        results = []
+        shared_accounts = set()
+        for r in resources:
+            accounts = self.manager.retry(
+                client.list_portfolio_access, PortfolioId=r['Id'], ignore_err_codes=(
+                    'ResourceNotFoundException',))
+            if accounts:
+                shared_accounts = set(accounts.get('AccountIds'))
+            delta_accounts = shared_accounts.difference(accounts)
+            if delta_accounts:
+                r[self.annotation_key] = list(delta_accounts)
+                results.append(r)
+        return results
+
+    def process(self, resources, event=None):
+        results = []
+        client = local_session(self.manager.session_factory).client('servicecatalog')
+        accounts = self.get_accounts()
+        results.extend(self.violation_checker(client, accounts, resources))
+        return results
+
+
+@CatalogPortfolio.action_registry.register('remove-shared-accounts')
+class RemoveSharedAccounts(BaseAction):
+    """Action to delete Portfolio share with other accounts
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: catalog-portfolio-delete-share
+                resource: aws.catalog-portfolio
+                filters:
+                  - type: cross-account
+                actions:
+                  - type: remove-shared-accounts
+                    accounts: matched
+    """
+
+    schema = type_schema(
+        'remove-shared-accounts',
+        accounts={'oneOf': [
+            {'enum': ['matched']},
+            {'type': 'array', 'items': {'type': 'string', 'minLength': 12, 'maxLength': 12}}]},
```

Comment:
Would this essentially serve as a blacklist of accounts? It feels oddly specific.

Reply:
I added this functionality in case the user has a static known account with which the portfolio was shared and wants to undo it

Reply:
Addressed